[Gekidou MM-44943] Add team picker to the search screen and results screens

jfrerich opened the pull request:
#### Summary
This PR adds the team picker icon in the Search Screen and Results.

#### Remaining Bugs
- [x] Joing a team from the sidebar works, but `onTeamAdded` in `add_team_slide_up.tsx` is getting call way to many times. (fixed in https://github.com/mattermost/mattermost-mobile/pull/6455/commits/7d8427ca97b70d197cb9fcf9dbb36fd1d484f0ce)
- [ ]  Calculate the DisplayTerm before storing in WDB

#### Search Screen
- [x] icon added Parallel to search options
- [x] recent searches will show team specific searches when the team is changed
- [x] Recent search title text includes the currently searching team

#### Team picker icon
- [x] shows the currently searching team icon or short team name text
- [x] tapping icon will open the `Select team search` bottomScreen

#### BottomScreen Team picker
- [x] changing the team will changes the searching team
- [x] when opening again, the latest selected team is selected in the menu
- [x] Leverages the existing `TeamList` from `@components/team_sidebar/add_team/team_list.tsx`
  - [ ] refactor to move `TeamList` into `@components/team_list`

#### Results 
- [x] team picker is shown in RHS of header menu for files and messages
- [x] will reload the searches for the currenly selected team
- [x] saves the current search term to the recent searches for that newly selected team

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-44943

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->

#### Screenshots
Search Screen
<img width="401" alt="image" src="https://user-images.githubusercontent.com/7575921/177061171-c3bbde53-7abb-43dd-a8e3-632c4fe91de6.png">

Results Screen (Messages tab)
<img width="392" alt="image" src="https://user-images.githubusercontent.com/7575921/177061200-b9dc8d4d-dc50-432c-b37d-7a168897607d.png">

Results Screen (Files tab)
<img width="341" alt="image" src="https://user-images.githubusercontent.com/7575921/177061238-ee050772-050b-46af-a06c-cc5928d5cbaa.png">

#### New bottom screen for picking search team
<img width="432" alt="image" src="https://user-images.githubusercontent.com/7575921/177061264-51c65023-671c-4ef7-bdd7-77110f9c91b3.png">


#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note

```

jfrerich commented on app/components/team_sidebar/add_team/add_team.tsx:
```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
This file was formerly the `.index.tsx` in the `add_team` directory.  The only change is to observe the `currentUserId` to pass to `add_team_slide_up.tsx`  Another option is to:
 
* keep `add_team.tsx` as the `index.tsx` 
* create subcomponent `add_team_slide_up/` with `.index.tsx` to get the current user id

jfrerich commented on app/components/team_sidebar/add_team/add_team_slide_up.tsx:
```diff
@@ -3,50 +3,112 @@
+                />
+            }
+            {!hasOtherTeams &&
+                <View style={styles.empty}>
```
This could also be added as a prop to `TeamList`...

larkox commented:
The tabs start to worry me a bit. Keep in mind that in some languages, the length of the text may be long. For example, in german you could have something like:
`Nachrichten (60) | Dateien (60)`

We may want to think how those may or may not apply some ellipsis.

larkox commented on app/screens/home/search/team_picker_icon/team_picker_icon.tsx:
```diff
@@ -0,0 +1,122 @@
+            theme,
+            title: intl.formatMessage({id: 'mobile.search.team.select', defaultMessage: 'Select a team to search'}),
+        });
+    }), [theme, teamId]);
```
`setTeamId` comes from the props, so it should be part of the dependencies. Same for teams.

jfrerich replied:
Added in https://github.com/mattermost/mattermost-mobile/pull/6455/commits/2e407ff5438f3dbfc25503aec1b02011f8107d42

matthewbirtch reviewed (changes requested):
@jfrerich a few things from me:

1. The size of the team selection icon is different in the default state vs the search results state. Is there a reason for that? If we need to go smaller, that's fine, but it should be consistent in both states. (see comparison below)

<img width="486" alt="image" src="https://user-images.githubusercontent.com/2040554/177341667-41a57411-8047-4d0b-a6ef-33043de70a4c.png">

2. When viewing results and then switching teams, the screen changes layout around and it feels unpolished. Are we able to keep the top bar in the collapsed state while this is all happening instead of jump around like this? 

<img src="https://user-images.githubusercontent.com/2040554/177342201-99a85240-a803-475b-be0f-1af969529047.gif" width="450">

3. Also, tapping the clear `x` on the search input does not reset and clear the entire search view. It seems to keep the results shown. It will only reset the view if I focus on the input again after clearing and hit `Cancel`. Can we make sure the entire search view gets reset when I hit the clear `x` icon within the search input? NOTE: I tested this is on iOS - not sure about Android

4. If I search initially in one team, then switch teams in the results view, then clear my search, the initial view should have the most recently team selected, but it doesn't.

5. Can we reduce the space between the items in the team selection bottom sheet list? It looks like a height is being set to 64 on the team_list_item container for some reason, but it should be 40. And then the vertical margin on these items should be 8.
<img width="410" alt="Screen Shot 2022-07-05 at 9 57 59 AM" src="https://user-images.githubusercontent.com/2040554/177345621-031e3279-3286-43fc-9d6e-208283f607ba.png">

7. Also, in the team selection bottom sheet, the rounded corner radius seem to be different between the fallback team icons and the image icons. Can we make them all 8 radius in this view?

8. The fallback team icons are not getting scaled appropriately. The letters are too large and getting cropped.
<img width="96" alt="image" src="https://user-images.githubusercontent.com/2040554/177347958-bc9f87be-915b-48fe-90d4-ad3f0a68e682.png">

9. The date divider line shows through the header on the left side when you scroll through results.
<img width="239" alt="Screen Shot 2022-07-05 at 10 13 43 AM" src="https://user-images.githubusercontent.com/2040554/177348592-1c33bf61-df5e-4cc4-8c01-6a566f19c5ae.png">

jfrerich commented on app/components/team_sidebar/add_team/add_team_slide_up.tsx:
```diff
@@ -3,50 +3,112 @@
     const onTeamAdded = useCallback(async (teamId: string) => {
-        await dismissBottomSheet();
-        handleTeamChange(serverUrl, teamId);
+        const {error} = await addUserToTeam(serverUrl, teamId, currentUserId);
```
This code originally resided inside the `TeamList` component, which I am now borrowing for the search team picker.  In order to make `TeamList` generic, I pulled it this code out and added the `onPress` callback prop to `TeamList`.  Does that answer your question?

jfrerich replied:
> `onTeamAdded` but `onPress`
 added in  onTeamAdded but onPress

jfrerich commented on app/screens/home/search/team_picker_icon/team_picker_icon.tsx:
```diff
@@ -0,0 +1,122 @@
+            snapPoints: [height, 10],
+            theme,
+            title: intl.formatMessage({id: 'mobile.search.team.select', defaultMessage: 'Select a team to search'}),
+        });
```
Yes, you are correct. `AddTeam` and this component, `TeamPickerIcon` both open the bottom sheet with content of `TeamList`, with some varying text. I created a `bottomSheetWithTeamList()` utils function.  This allowed sharing all the height and snap calculations.

See https://github.com/mattermost/mattermost-mobile/pull/6455/commits/6c1553a0c12c50f43dd2676c211b584f11995df8 for how I implemented the change.

jfrerich commented on app/screens/home/search/results/header.tsx:
```diff
@@ -92,7 +98,7 @@ const Header = ({
             snapPoints: [700, 10],
             theme,
-            title: intl.formatMessage({id: 'mobile.add_team.join_team', defaultMessage: 'Join Another Team'}),
+            title: '',
```
I didn't understand it's usage for tables. I added it back. Also, I noticed that unless we have a different title in the `renderContent` results, we could just have bottomSheet determine whether to show the title in the tablet title or add the renderContent() title automatically.  We wouldn't need to pass to both `renderContend()` and `bottomSheet()`

I added the title back here https://github.com/mattermost/mattermost-mobile/pull/6455/commits/e28ed0979260d5160cb966c56487b9337fd92ed6

jfrerich replied:
ticket created.  https://mattermost.atlassian.net/browse/MM-45570

jfrerich commented on app/components/team_sidebar/add_team/team_list_item/team_list_item.tsx:
```diff
@@ -88,6 +88,15 @@ export default function TeamListItem({team, currentUserId, textColor, iconTextCo
                 >
                     {displayName}
                 </Text>
+                {selectedTeamId && (team.id === selectedTeamId) &&
```
correct. I made the change and verified.

jfrerich commented:
@matthewbirtch, can you verify the folllowing?

> 7. Also, in the team selection bottom sheet, the rounded corner radius seem to be different between the fallback team icons and the image icons. Can we make them all 8 radius in this view?
> 8. The fallback team icons are not getting scaled appropriately. The letters are too large and getting cropped.

From this commit https://github.com/mattermost/mattermost-mobile/pull/6455/commits/a91d0a7194d551aba3a1cb735cef741dbec9ecbf

- now all team_icons are all radius of 8. Includes the following icons:
  - team picker icon in the search screen header
    - each team icon in the search screen bottom sheet team list
  - team picker icon in the search results screen header
  - each team icon in the home screen team side bar list
    - each team icon in the join new team bottom sheet team list (from home screen)

jfrerich commented:
> 9\. The date divider line shows through the header on the left side when you scroll through results.

This can also be verified as fixed in the last commit https://github.com/mattermost/mattermost-mobile/pull/6455/commits/44d162f3fbbd848ddab56891d973c17c150cf330

matthewbirtch commented:
> * Also, in the team selection bottom sheet, the rounded corner radius seem to be different between the fallback team icons and the image icons. Can we make them all 8 radius in this view?

Corner radius is good now for all team icons.

> * The fallback team icons are not getting scaled appropriately. The letters are too large and getting cropped.

The font size of the letters in the fallback team icon that appears in the top-right of the search view should be smaller in that context. Can we adjust the font-size to 16 in this specific case?

matthewbirtch commented:
> > 9. The date divider line shows through the header on the left side when you scroll through results.
> 
> This can also be verified as fixed in the last commit [44d162f](https://github.com/mattermost/mattermost-mobile/commit/44d162f3fbbd848ddab56891d973c17c150cf330)

@jfrerich this removed the rounded corners of the main content area now though
<img width="477" alt="image" src="https://user-images.githubusercontent.com/2040554/177447658-f1aa19c3-d5e3-418a-ba90-9193fdd83381.png">

jfrerich commented:
> > * The fallback team icons are not getting scaled appropriately. The letters are too large and getting cropped.
> 
> The font size of the letters in the fallback team icon that appears in the top-right of the search view should be smaller in that context. Can we adjust the font-size to 16 in this specific case?

done.  

I added a smallText Prop to the TeamIcon component.  setting to true will set the use typography with size value of 200 instead of the default of 400.

jfrerich commented on app/screens/home/search/search.tsx:
```diff
@@ -97,15 +98,15 @@ const SearchScreen = ({teamId}: Props) => {
         setFilter(FileFilters.ALL);
         setLastSearchedValue(term);
-        addSearchToTeamSearchHistory(serverUrl, teamId, term);
+        addSearchToTeamSearchHistory(serverUrl, searchTeamId, term);
```
Yes.  Thanks for catching this.

jfrerich commented:
> > I added a smallText Prop to the TeamIcon component. setting to true will set the use typography with size value of 200 instead of the default of 400.
> 
> This worked to change the size, but unfortunately the font family used for this should always be `Metropolis Semibold`, so it probably won't work with the `typography` helper.

Just to make sure.  We want the  team icon font family to have `Metropolis Semibold` for the team picker Icons in the search and results header and also the team list in the team picker? But we want the team icons in the sidebar to remain as the were `...typography('Heading', 400, 'SemiBold')`?

jfrerich commented:
> > > 9. The date divider line shows through the header on the left side when you scroll through results.
> > 
> > 
> > This can also be verified as fixed in the last commit [44d162f](https://github.com/mattermost/mattermost-mobile/commit/44d162f3fbbd848ddab56891d973c17c150cf330)
> 
> @jfrerich this removed the rounded corners of the main content area now though <img alt="image" width="477" src="https://user-images.githubusercontent.com/2040554/177447658-f1aa19c3-d5e3-418a-ba90-9193fdd83381.png">

@matthewbirtch, please check the latest commit to see if this works for you now.

larkox commented on app/screens/home/search/search.tsx:
```diff
@@ -83,7 +84,7 @@ const SearchScreen = ({teamId}: Props) => {
     const onSubmit = useCallback(() => {
-        handleSearch(searchValue);
+        handleSearch(searchTeamId, searchValue);
     }, [searchValue]);
```
This seems to depend also on `searchTeamId`

larkox reviewed (changes requested):
Some minor changes, and should be ready to go.

matthewbirtch commented:
> Just to make sure. We want the team icon font family to have `Metropolis Semibold` for the team picker Icons in the search and results header and also the team list in the team picker? But we want the team icons in the sidebar to remain as the were `...typography('Heading', 400, 'SemiBold')`?

The team icon fallback should always use `Metropolis Semibold` regardless of size. Using `...typography('Heading', 400, 'SemiBold')` works in the sidebar because that style uses Metropolis.

matthewbirtch commented:
> > > > 9. The date divider line shows through the header on the left side when you scroll through results.
> > > 
> > > 
> > > This can also be verified as fixed in the last commit [44d162f](https://github.com/mattermost/mattermost-mobile/commit/44d162f3fbbd848ddab56891d973c17c150cf330)
> > 
> > 
> > @jfrerich this removed the rounded corners of the main content area now though <img alt="image" width="477" src="https://user-images.githubusercontent.com/2040554/177447658-f1aa19c3-d5e3-418a-ba90-9193fdd83381.png">
> 
> @matthewbirtch, please check the latest commit to see if this works for you now.

@jfrerich, sorry it is still not quite right. I'm seeing the the content show through the bottom of the header area now
<img width="490" alt="Screen Shot 2022-07-07 at 9 05 39 AM" src="https://user-images.githubusercontent.com/2040554/177780724-10db1d2c-34e5-4d04-9ac5-e13f6acba21e.png">

jfrerich commented:
> > Just to make sure. We want the team icon font family to have `Metropolis Semibold` for the team picker Icons in the search and results header and also the team list in the team picker? But we want the team icons in the sidebar to remain as the were `...typography('Heading', 400, 'SemiBold')`?
> 
> The team icon fallback should always use `Metropolis Semibold` regardless of size. Using `...typography('Heading', 400, 'SemiBold')` works in the sidebar because that style uses Metropolis.

Ok @matthewbirtch, I hope this last commit fixes the issue.  I do use the typography help for getting the size. I then modify the at return value to always use `Metropolis Semibold`. please verify

larkox commented on app/components/team_sidebar/team_list/team_item/team_icon.tsx:
```diff
@@ -101,9 +101,12 @@ export default function TeamIcon({
     }, [styles, backgroundColor, selected, nameOnly]);
 
+    const textTypography = typography('Heading', smallText ? 200 : 400, 'SemiBold');
+    textTypography.fontFamily = 'Metropolis-SemiBold';
```
This should not be needed, but I see that for size 400 the SemiBold style is not added... Can you file a ticket to investigate?

enahum commented:
Is anyone checking that all this changes are not affecting ffunctionality and UI/UX in the other places this modified components are used and all cases are covered??

I must say that I just took a quick look and I'm not entirely sure this has zero impact, just saying.. during the reviews do not focus only in the immediate affected areas but do look at everything else.

avinashlng1080 commented on app/components/team_sidebar/team_list/team_item/team_icon.tsx:
```diff
@@ -100,11 +101,15 @@ export default function TeamIcon({
+
     let teamIconContent;
     if (nameOnly) {
+        const textStyle = [styles.text, textTypography];
```
```suggestion
      const textStyle = [
            styles.text,
            textTypography,
            textColor && {color: textColor},
        ];
```

avinashlng1080 commented on app/components/team_sidebar/team_list/team_item/team_icon.tsx:
```diff
@@ -100,11 +101,15 @@ export default function TeamIcon({
         teamIconContent = (
             <Text
-                style={textColor ? [styles.text, {color: textColor}] : styles.text}
+                style={textColor ? [textStyle, {color: textColor}] : textStyle}
```
```suggestion
                 style={textStyle}
```

jfrerich replied:
much cleaner

avinashlng1080 commented on app/screens/home/search/modifiers/index.tsx:
```diff
@@ -135,11 +123,18 @@ const SearchModifiers = ({searchValue, setSearchValue}: Props) => {
+                    defaultMessage='Search options'
+                />
+                <TeamPickerIcon
+                    size={32}
```
magic number - you can move it into a constant

avinashlng1080 commented on app/screens/home/search/team_picker_icon/search_team_slideup.tsx:
```diff
@@ -0,0 +1,47 @@
+}
+
+export default function SelectTeamSlideUp({teams, title, setTeamId, teamId}: Props) {
+    const testID = 'team_sidebar.add_team_slide_up';
```
no need for this constant - use a string value at line 34 directly

avinashlng1080 commented on app/screens/home/search/team_picker_icon/search_team_slideup.tsx:
```diff
@@ -0,0 +1,47 @@
+
+    const onPress = useCallback((newTeamId: string) => {
+        setTeamId(newTeamId);
+        dismissBottomSheet();
```
Perhaps you need to `await dismissBottomSheet();` first and then call `setTeamId(newTeamId)`

jfrerich replied:
If I `await dismissBottomSheet()` first, the user sees two things 

1) The team icon change 
2) the recent changes section flicker from the old team values to the new.  

As it currently is, the change the team, and see the updated underlying data after the dismissal, as it has already changed via `setTeamId`

avinashlng1080 commented on app/screens/home/search/team_picker_icon/search_team_slideup.tsx:
```diff
@@ -0,0 +1,47 @@
+            testID={testID}
+            title={title}
+        >
+            {teams.length &&
```
should this component at all return null if the teams length is 0?

jfrerich replied:
hmm.. That's an interesting question.  I believe a user must belong to at least one team, so teams.length should never be 0. The question for @matthewbirtch, do we want to disable the team picker if the user only belongs to one team?  I'm going to create this as a Jira ticket for now. https://mattermost.atlassian.net/browse/MM-45623

matthewbirtch replied:
It is possible (although unlikely) that a user is not a member of any teams. However, I don't believe the bottom tab bar (and therefore the search view) is accessible if the user is not a member of a team.

avinashlng1080 commented on app/screens/home/search/team_picker_icon/team_picker_icon.tsx:
```diff
@@ -0,0 +1,114 @@
+    setTeamId: (id: string) => void;
+    teamId: string;
+}
+const TeamPickerIcon = ({size = 24, divider = false, setTeamId, teams, teamId}: Props) => {
```
24 - avoid magic number

avinashlng1080 commented on app/screens/home/search/team_picker_icon/team_picker_icon.tsx:
```diff
@@ -0,0 +1,114 @@
+                    type='opacity'
+                    testID={selectedTeam.id}
+                >
+                    <View style={[styles.teamContainer, divider ? styles.border : null]}>
```
```suggestion
                       <View style={[styles.teamContainer, divider && styles.border]}>
```

jfrerich commented on app/actions/remote/search.ts:
```diff
@@ -40,7 +40,7 @@ export async function fetchRecentMentions(serverUrl: string): Promise<PostSearch
         }
         const terms = currentUser.userMentionKeys.map(({key}) => key).join(' ').trim() + ' ';
-        const results = await searchPosts(serverUrl, {terms, is_or_search: true});
+        const results = await searchPosts(serverUrl, '', {terms, is_or_search: true});
```
That is correct.  On post search, I limit to the teamId, but it looks like that wanted all mentions across all teams.

matthewbirtch commented:
> Is anyone checking that all this changes are not affecting ffunctionality and UI/UX in the other places this modified components are used and all cases are covered??
> 
> I must say that I just took a quick look and I'm not entirely sure this has zero impact, just saying.. during the reviews do not focus only in the immediate affected areas but do look at everything else.

I have been looking at the team icons and bottom sheets for team selection in the other places and making sure it's ok. Haven't seen any issues elsewhere

matthewbirtch commented:
> Ok @matthewbirtch, I hope this last commit fixes the issue. I do use the typography help for getting the size. I then modify the at return value to always use `Metropolis Semibold`. please verify

Verified that the fallback icons look good now @jfrerich

jfrerich commented:
Thanks for the feedback, Matt!

I have not forgot about your remaining navigation flakiness issues. I was getting lots of code PR feedback and wanted to catch those guys up.  In the event that I we don't merge these before I leave on Monday, I wantede them to get further along.  

I'm working on navigation now.

matthewbirtch commented:
> Thanks for the feedback, Matt!
> 
> I have not forgot about your remaining navigation flakiness issues. I was getting lots of code PR feedback and wanted to catch those guys up. In the event that I we don't merge these before I leave on Monday, I wantede them to get further along.
> 
> I'm working on navigation now.

Oh shoot....I kind of forgot about that (got lost with all the conversation. I don't know how to 'unapprove'

matthewbirtch reviewed (changes requested):
Sorry, forgot the flakiness with team switching on results @jfrerich and the top nav bar. Taking back the approval :)

jfrerich commented:
You just press the recycle looking button. <img width="336" alt="image" src="https://user-images.githubusercontent.com/7575921/178116099-1d81131a-a155-4344-813f-0f1b74fdcf03.png">

jfrerich commented:
@matthewbirtch, I think I finally resolved issue 9. To recap, the initial issue was the results showing on through the header with messages and files tabs.  I am not sure why I implemented the way I did, but I had a small divider view with height of 1 that was added. I removed that and just set the bottom border width and color as part of the header container.

I am still trying to work on your issues 2 and 3, although I might have to create tickets for these. I'll let you know by the end of today.

jfrerich commented:
I created https://mattermost.atlassian.net/browse/MM-45625 which details the remaining navigation type issues found in this ticket. If you find any more, can you please add them to the ticket.  We could then get these merged to get the functionality in and then tweak the remaining issues.

matthewbirtch reviewed (approved):
Thanks @jfrerich, I'm good with moving those issues out to their own ticket so we can get this merged sooner.

avinashlng1080 reviewed (approved):
LGTM.
